### PR TITLE
Feature/prune head

### LINKED ISSUE #812: add a `nn_prune_head()` utility function
Hello,

It took me a while to find an good way to prune head of a pretrained nn_module, to turn it into an embedding module.
Finally 
```r
pruned_model <- torch::nn_sequential(!!!model$children[1:(length(model$children)-1)])
```
did it.

I think it would be convenient to have this as an utility function, either with explicit parameter `module=1L` either with an heuristic to prune all fully-connected / linear / batch_norm / *elu layers from the tail.

### PR DESCRIPTION
**Pull request by cregouby**
fix #812 

I've a few test that pass on mlverse packages and friends, but I'm not sure if {torch} is the right location to host them. I include them for reference (and or later on dispatch) 
## {tochvision} model pruning 

```{r}
test_that("we can prune head of torchvision resnet34 moels", {
  resnet34 <- model_resnet34(pretrained=TRUE)

  expect_error(prune <- nn_prune_head(resnet34, 1), NA)
  # expect_true(inherits(prune, "nn_sequential"))
  expect_equal(length(prune), 9)
  expect_true(inherits(prune[[length(prune)]], "nn_adaptive_avg_pool2d"))
  
  input <- torch::torch_randn(1, 3, 256, 256)
  out <- prune(input)
  expect_tensor_shape(out, c(1, 512, 1, 1))

})

test_that("we can prune head of torchvision resnet50 moels", {
  resnet50 <- model_resnet50(pretrained=TRUE)

  expect_error(prune <- nn_prune_head(resnet50, 1), NA)
  expect_true(inherits(prune, "nn_sequential"))
  expect_equal(length(prune), 9)
  expect_true(inherits(prune[[length(prune)]], "nn_adaptive_avg_pool2d"))
    
  input <- torch::torch_randn(1, 3, 256, 256)
  out <- prune(input)
  expect_tensor_shape(out, c(1, 2048, 1, 1))


})

test_that("we can prune head of torchvision resnext101 moels", {
  resnext101 <- model_resnext101_32x8d(pretrained=TRUE)

  expect_error(prune <- torch:::nn_prune_head(resnext101, 1), NA)
  expect_true(inherits(prune, "nn_sequential"))
  expect_equal(length(prune), 9)
  expect_true(inherits(prune[[length(prune)]], "nn_adaptive_avg_pool2d"))
    
  input <- torch::torch_randn(1, 3, 256, 256)
  out <- prune(input)
  expect_tensor_shape(out, c(1, 2048, 1, 1))


})
test_that("we can prune head of torchvision mobilenetv2 moels", {
  mobilenet <- model_mobilenet_v2(pretrained=TRUE)

  expect_error(prune <- nn_prune_head(mobilenet, 1), NA)
  expect_true(inherits(prune, "nn_sequential"))
  expect_equal(length(prune), 1)
  expect_true(inherits(prune[[length(prune)]], "nn_sequential"))
    
  input <- torch::torch_randn(1, 3, 256, 256)
  out <- prune(input)
  expect_tensor_shape(out, c(1, 1280, 8, 8))
})
```
## {torchaudio} model pruning

```{r}
test_that("we can prune head of torchaudio melresnet models", {
  melresnet  <-  model_melresnet()
  
  expect_error(prune <- torch:::nn_prune_head(melresnet[[1]], 1), NA)
  expect_true(inherits(prune, "nn_sequential"))
  # expect_equal(length(prune), 9)
  expect_true(inherits(prune[[length(prune)]], "ResBlock"))
  
  input  <-  torch::torch_rand(10, 128, 512)  
  output  <-  prune(input)  # shape: (10, 128, 508)
  expect_tensor_shape(output, c(10, 128, 508))
})  
```
# {torchtransformer} model pruning
```{r}
emb_size <- 128L
mpe <- 512L
n_head <- 4L
n_layer <- 6L
vocab_size <- 30522L
n_inputs <- 2
n_token_max <- 128L
t_ids <- matrix(
  sample(
    2:vocab_size,
    size = n_token_max * n_inputs,
    replace = TRUE
  ),
  nrow = n_token_max, ncol = n_inputs
)
ttype_ids <- matrix(
  rep(1L, n_token_max * n_inputs),
  nrow = n_token_max, ncol = n_inputs
)

test_that("we can prune head of torchtransformer  model", {
  model <- model_bert(
    embedding_size = emb_size,
    n_layer = n_layer,
    n_head = n_head,
    max_position_embeddings = mpe,
    vocab_size = vocab_size
  )
  expect_error(prune <- nn_prune_head(model), NA)
  output  <-  prune(
    torch::torch_tensor(t_ids),
    torch::torch_tensor(ttype_ids)
    
)
  expect_tensor_shape(output, c(128, 2, 128))

})
```

### CODE REVIEW

File: R/nn.R [hunk 1]
```diff
@@ -637,6 +637,56 @@ length.nn_sequential <- function(x) {
...
+  UseMethod("nn_prune_head")
+}
+
+#' @export
+nn_prune_head.nn_sequential <- function(x, head_size=1L ) {
+  nn_sequential(!!!x$children[1:(length(x)-head_size)])
+}
+
+#' @export
+nn_prune_head.nn_module <- module <- nn_module(
```
Comment (dfalbel) [inline R/nn.R:668]:
```suggestion
nn_prune_head.nn_module <- nn_module(
```

File: R/nn.R [hunk 2]
```diff
@@ -637,6 +637,56 @@ length.nn_sequential <- function(x) {
...
+    classname = "nn_sequential",
+    initialize = function(x, head_size=1L) {
+      modules <- rlang::list2(!!!x$children[1:(length(x$children)-head_size)])
+      mod_names <- names(modules) 
+      for (i in seq_along(modules)) {
+        self$add_module(name = mod_names[i], module = modules[[i]])
+      }
+    },
+    forward = function(...) {
+      i <- 1
```
Comment (dfalbel) [inline R/nn.R:678]: I don't see where `i` is updated here. Shouldn't it change in the loop?

**Review by dfalbel — COMMENTED** [review]
Thanks @cregouby ! This is very nice :) 
Just added 2 comments.

**Comment by dfalbel** [comment]
Regarding the extra tests, let's wait for the version containing `nn_prune_head()` makes it to CRAN and we can add it to the respective packages.

File: R/nn.R (hunk 2 above)
Comment (cregouby) [reply R/nn.R]: you're absolutely right. my bad, sorry. I've turned it into an explicit flag.

**Comment by dfalbel** [comment]
Thank you @cregouby !

**Comment by cregouby** [comment]
My pleasure !

**Pull request merged**